### PR TITLE
ObjectEntryDescriptor Serializer Trait

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Ended up being quite a mess. Not fully tested.

### CODE REVIEW

File: src/openrct2/object/Object.h [hunk 1]
```diff
@@ -139,11 +139,23 @@ struct ObjectEntryDescriptor
     ObjectGeneration Generation;
     union
     {
+        std::string Identifier; // For JSON objects
```
Comment (duncanspumpkin) [inline src/openrct2/object/Object.h:142]: Might just break this union into two fields. The Union is causing a good few issues.

File: src/openrct2/core/DataSerialiserTraits.h [hunk 2]
```diff
@@ -730,3 +731,46 @@ template<> struct DataSerializerTraits_t<rct_vehicle_colour>
...
+            std::string id;
+            DataSerializerTraits<std::string> s;
+            s.decode(stream, id);
+            val = ObjectEntryDescriptor(id);
+        }
+    }
+    static void log(OpenRCT2::IStream* stream, const ObjectEntryDescriptor& val)
+    {
+        char msg[128] = {};
+        snprintf(msg, sizeof(msg), "ObjectEntryDescription (Generation = %d)", val.Generation);
```
Comment (Gymnasiast) [inline src/openrct2/core/DataSerialiserTraits.h:773]:
```suggestion
        snprintf(msg, sizeof(msg), "ObjectEntryDescriptor (Generation = %d)", val.Generation);
```

File: src/openrct2/object/Object.h [hunk 3]
```diff
@@ -137,60 +152,30 @@ enum class ObjectGeneration : uint8_t
...
+        , Entry()
     {
-        *this = entry;
     }
 
     explicit ObjectEntryDescriptor(const rct_object_entry& newEntry)
     {
         Generation = ObjectGeneration::DAT;
         Entry = newEntry;
+        Identifier = std::string(newEntry.GetName());
```
Comment (Gymnasiast) [inline src/openrct2/object/Object.h:169]: Not entirely sure about this. I'd prefer having `Identifier` empty if there isn't one. (Also see the conversation in #plug-in with Bassie.)

File: src/openrct2/core/DataSerialiserTraits.h [hunk 4]
```diff
@@ -730,3 +731,46 @@ template<> struct DataSerializerTraits_t<rct_vehicle_colour>
...
+            std::string id;
+            DataSerializerTraits<std::string> s;
+            s.decode(stream, id);
+            val = ObjectEntryDescriptor(id);
+        }
+    }
+    static void log(OpenRCT2::IStream* stream, const ObjectEntryDescriptor& val)
+    {
+        char msg[128] = {};
+        snprintf(msg, sizeof(msg), "ObjectEntryDescriptor (Generation = %d)", static_cast<int32_t>(val.Generation));
```
Comment (Gymnasiast) [inline src/openrct2/core/DataSerialiserTraits.h:773]: This won't log the identifier/DAT name, which is the most interesting bit.

This _might_ be fine, I don't know how important this logging function is.

**Review by Gymnasiast — APPROVED** [review]
LGTM. I have one line note, but I'll let you decide if that needs addressing or not.

File: src/openrct2/core/DataSerialiserTraits.h (hunk 4 above)
Comment (duncanspumpkin) [reply src/openrct2/core/DataSerialiserTraits.h:773]: Loging function is only used in network code as far as i am aware so you can't actually trigger this.

**Pull request merged**